backend/cfg/cfg_dataflow.ml

```diff
@@ -132,3 +132,192 @@ module Forward (D : Domain) (T : Transfer with type domain = D.t) :
+    | Tailcall _ -> true
+    | Call_no_return _ -> true
+
+  let create : Cfg.t -> init:domain -> map * WorkSet.t ref =
+   fun cfg ~init ->
+    let map = Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks) in
+    let set = ref WorkSet.empty in
+    let value = init in
+    Cfg.iter_blocks cfg ~f:(fun label block ->
+        if is_exit_block block || true
+           (* CR xclerc for xclerc: remove `|| true` *)
+        then begin
+          Label.Tbl.replace map label value;
+          set := WorkSet.add { WorkSetElement.label; value } !set
+        end);
+    map, set
+
+  let remove_and_return :
+      Cfg.t -> WorkSet.t ref -> WorkSetElement.t * Cfg.basic_block =
+   fun cfg set ->
+    let element = WorkSet.choose !set in
+    set := WorkSet.remove element !set;
+    element, Cfg.get_block_exn cfg element.label
+
+  let run :
+      Cfg.t -> ?max_iteration:int -> init:domain -> unit -> (map, map) Result.t
+      =
+   fun cfg ?(max_iteration = max_int) ~init () ->
+    let res, work_set = create cfg ~init in
+    let iteration = ref 0 in
+    (* note: `handler_map` contains the value at the *start* of the block. *)
+    let handler_map : D.t Label.Tbl.t =
+      Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks)
+    in
+    while (not (WorkSet.is_empty !work_set)) && !iteration < max_iteration do
+      incr iteration;
+      let element, block = remove_and_return cfg work_set in
+      let exn : domain =
+        Option.map
+          (fun exceptional_successor ->
+            Label.Tbl.find_opt handler_map exceptional_successor)
+          block.exn
+        |> Option.join
+        |> Option.value ~default:D.bot
+      in
+      let value = transfer_block element.value ~exn block in
+      if block.is_trap_handler
+      then begin
+        let old_value =
+          Option.value
+            (Label.Tbl.find_opt handler_map block.start)
+            ~default:D.bot
+        in
+        let new_value = T.exception_ value in
+        if not (D.less_equal new_value old_value)
+        then begin
+          Label.Tbl.replace handler_map block.start new_value;
+          List.iter
+            (fun predecessor_label ->
+              let current_value =
+                Option.value
+                  (Label.Tbl.find_opt res predecessor_label)
+                  ~default:D.bot
+              in
+              work_set
+                := WorkSet.add
+                     { WorkSetElement.label = predecessor_label;
+                       value = current_value
+                     }
+                     !work_set)
+            (Cfg.predecessor_labels block)
+        end
+      end
+      else
+        List.iter
+          (fun predecessor_label ->
+            let old_value =
+              Option.value
+                (Label.Tbl.find_opt res predecessor_label)
+                ~default:D.bot
+            in
+            let new_value = D.join old_value value in
+            if not (D.less_equal new_value old_value)
```

Comment:
Similarly, forward analysis should use `less_equal` not compare  for this.

Reply:
Yes, I think it is implied by what I said "offline" the other day.